DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd.hpp

```diff
@@ -0,0 +1,273 @@
+    std::terminate();
+  }
+  if (v_matrix_device == nullptr) {
+    std::cerr << "Error when allocating the V matrix." << std::endl;
+    std::terminate();
+  }
+  if (r_device == nullptr) {
+    std::cerr << "Error when allocating the R matrix." << std::endl;
+    std::terminate();
+  }
+
+  // copy input matrix to device
+  q.memcpy(input_matrix_device, a_matrix.data(),
+           kInputMatrixSize * sizeof(float))
+      .wait();
+
+  // read input matrix from DDR
+  auto ddr_write_event = q.submit([&](sycl::handler &h) {
+    h.single_task<IDInputMatrixFromDDRToLocalMem>(
+        [=]() [[intel::kernel_args_restrict]] {
+          MatrixReadFromDDRToPipeByBlocks<
+              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe>(
+              input_matrix_device, 1, repetitions);
+        });
+  });
+
+  // Compute the covariance matrix
+  q.single_task<IDCovarianceMatrixComputation>(
+      fpga_linalg::StreamingNStdCovarianceMatrix<
+          T, rows, cols, kNumElementsPerDDRBurst, InputMatrixPipe,
+          CovarianceMatrixPipe>());
+
+  // Compute the Eigen values and Eigen vectors
+  q.single_task<IDEigenValuesAndVectorsComputation>(
+      fpga_linalg::StreamingEigen<T, cols, k_fixed_iteration,
+                                  kNumElementsPerDDRBurst, k_zero_threshold_1e,
+                                  CovarianceMatrixPipe, EigenValuesPipe,
+                                  EigenVectorsPipe, RankDeficientFlagPipe>());
+
+  // read input matrix from DDR (again)
+  auto ddr_write_event_post = q.submit([&](sycl::handler &h) {
+    h.single_task<IDInputMatrixFromDDRToLocalMemPostProcess>(
+        [=]() [[intel::kernel_args_restrict]] {
+          MatrixReadFromDDRToPipeByBlocks<
+              T, cols, rows, kNumElementsPerDDRBurst, InputMatrixPipe2>(
+              input_matrix_device, 1, repetitions);
+        });
+  });
+
+  // Post process. Using eigen values and eigen vector to produce U, S, V matrix
+  q.single_task<IDUSVFromEigens>(
+      fpga_svd::PostProcess<T, false, rows, cols, kNumElementsPerDDRBurst,
+                            InputMatrixPipe2, EigenValuesPipe, EigenVectorsPipe,
+                            UTempMatrixPipe, SMatrixPipe, VMatrixPipe>{});
+
+  // Orthogonalizing the U matrix by running it through QRD
+  q.single_task<IDOrthogonalizingU>(
+      fpga_linalg::StreamingQRD<T, false, rows, rows, k_raw_latency,
+                                kNumElementsPerDDRBurst, UTempMatrixPipe,
+                                UMatrixPipe, RMatrixPipe>());
+
+  // Write the rank deficient flag from local memory to FPGA DDR
+  auto rank_deficient_flag_event =
+      q.single_task<IDRankDeficientFlagFromLocalMemToDDR>(
+          [=]() [[intel::kernel_args_restrict]] {
+            VectorReadPipeToDDR<ac_int<1, false>, 1, RankDeficientFlagPipe>(
+                rank_deficient_flag_device, 1, repetitions);
+          });
+
+  // collect R matrix from the orthogonalizing kernel
```

Comment:
Is there any reason why we would want to output the R matrix? or can we just read from the pipe and do nothing with it to save memory bandwidth?